### DISCUSSION TITLE
A proposition to include built-in keywords to perform object mapping in C#

### DISCUSSION THREAD

**Discussion by bdicroce** [dotnet/csharplang#7448] [Language Ideas]
I'm not even sure if this idea makes sense, because I haven't seen it in other languages. Anyhow, it's quite a simple proposal: include built-in keywords to perform object mapping in C#.

I currently use the AutoMapper library (which is fantastic!), but part of me wish that this kind of functionality would be built-in the language.

I'm including a screenshot of what the syntax might look like (it's not a detailed as it should probably be, because I'm writing this during my lunch hour). The idea would be to have a new `map` keyword that will do the magic behind the scenes.

![carbon](https://github.com/dotnet/csharplang/assets/16505369/ab97ce26-d510-4a1e-8ef1-5bea8c297d35)

**Comment by HaloFour** [comment]
IMO mapping of entities is loaded with policy that is best handled outside of the language through tooling, libraries like AutoMapper or through source generator that can emit glue code as a part of compilation.

**Comment by spydacarnage** [comment]
As somebody who has written such a source generator, I can assure you it isn't as simple as a keyword.
Unless you completely ignore any properties which aren't _exactly_ the same (which makes it a lot less useful), the mapper would need to make too many decisions as to how to map things to be viable.

I have multiple attributes to help coerce the data in the correct way, and while it is very good at what it does, even now it still probably has a few holes that I haven't filled yet.
